src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,622 @@ end
+    r1 = ocr1.row
+    c11, c12 = ocr1.cols
+    r2 = ocr2.row
+    c21, c22 = ocr2.cols
+
+    return @inbounds isequal(c11[r1], c21[r2]) && isequal(c12[r1], c22[r2])
+end
+
+@inline function Base.isequal(ocr1::OnColRow{<:NTuple{3,AbstractVector}}, ocr2::OnColRow{<:NTuple{3,AbstractVector}})
+    r1 = ocr1.row
+    c11, c12, c13 = ocr1.cols
+    r2 = ocr2.row
+    c21, c22, c23 = ocr2.cols
+
+    return @inbounds isequal(c11[r1], c21[r2]) &&
+                     isequal(c12[r1], c22[r2]) && isequal(c13[r1], c23[r2])
+end
+
+@inline function Base.isequal(ocr1::OnColRow{<:NTuple{N,AbstractVector}}, ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {N}
+    r1 = ocr1.row
+    cols1 = ocr1.cols
+    r2 = ocr2.row
+    cols2 = ocr2.cols
+
+    @inbounds for i in 1:N
+        isequal(cols1[i][r1], cols2[i][r2]) || return false
+    end
+    return true
+end
+
+@inline function Base.isless(ocr1::OnColRow{<:NTuple{2, AbstractVector}}, ocr2::OnColRow{<:NTuple{2, AbstractVector}})
+    r1 = ocr1.row
+    c11, c12 = ocr1.cols
+    r2 = ocr2.row
+    c21, c22 = ocr2.cols
+
+    c11r = @inbounds c11[r1]
+    c12r = @inbounds c12[r1]
+    c21r = @inbounds c21[r2]
+    c22r = @inbounds c22[r2]
+
+    isless(c11r, c21r) && return true
+    isequal(c11r, c21r) || return false
+    return isless(c12r, c22r)
+end
+
+@inline function Base.isless(ocr1::OnColRow{<:NTuple{3,AbstractVector}}, ocr2::OnColRow{<:NTuple{3,AbstractVector}})
+    r1 = ocr1.row
+    c11, c12, c13 = ocr1.cols
+    r2 = ocr2.row
+    c21, c22, c23 = ocr2.cols
+
+    c11r = @inbounds c11[r1]
+    c12r = @inbounds c12[r1]
+    c13r = @inbounds c13[r1]
+    c21r = @inbounds c21[r2]
+    c22r = @inbounds c22[r2]
+    c23r = @inbounds c23[r2]
+
+    isless(c11r, c21r) && return true
+    isequal(c11r, c21r) || return false
+    isless(c12r, c22r) && return true
+    isequal(c12r, c22r) || return false
+    return isless(c13r, c23r)
+end
+
+@inline function Base.isless(ocr1::OnColRow{<:NTuple{N,AbstractVector}}, ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {T1, T2, N}
+    r1 = ocr1.row
+    cols1 = ocr1.cols
+    r2 = ocr2.row
+    cols2 = ocr2.cols
+
+    lastcols1 = @inbounds cols1[1][r1]
+    lastcols2 = @inbounds cols2[1][r2]
+    isless(lastcols1, lastcols2) && return true
+    @inbounds for i in 2:N
+        isequal(lastcols1, lastcols2) || return false
+        lastcols1 = cols1[i][r1]
+        lastcols2 = cols2[i][r2]
+        isless(lastcols1, lastcols2) && return true
+    end
+    return false
+end
+
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = OnCol(cs...)
+
+# Return if it is allowed to use refpool instead of the original array for joining.
+# There are multiple conditions that must be met to allow for this.
+# If it is allowed we are sure that nothing can be used as a sentinel
```

Comment:
Below you use `missing` as a sentinel, right? `nothing` indeed sounds more appropriate here (even if that doesn't make a difference).

Reply:
it makes a difference. If I use `missing` as a sentinel below I can use a fast path where I allow `AbstractVector{<:Union{Missing, Integer}}`. And I prefer to allow `Missing` rather than `Nothing` here, as users are more likely to pass just normal vectors of integers mixed with `missing` rather than integers mixed with `nothing`.